group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/HdrHistogram.java

```diff
@@ -70,19 +69,20 @@ public HdrHistogram(
     ) {
         this.maxSnapshotAgeMs = maxSnapshotAgeMs;
         recorder = new Recorder(highestTrackableValue, numberOfSignificantValueDigits);
-        this.timestampedHistogramSnapshot = new AtomicReference<>(new Timestamped<>(0, null));
+        this.timestampedHistogramSnapshot = new Timestamped<>(0, null);
     }
 
     private Histogram latestHistogram(long now) {
-        Timestamped<Histogram> latest = timestampedHistogramSnapshot.get();
-        while (now - latest.timestamp > maxSnapshotAgeMs) {
-            Histogram currentSnapshot = recorder.getIntervalHistogram();
-            boolean updatedLatest = timestampedHistogramSnapshot.compareAndSet(
-                latest, new Timestamped<>(now, currentSnapshot));
-
-            latest = timestampedHistogramSnapshot.get();
-            if (updatedLatest) {
-                break;
```

Comment:
I was able to reproduce the bug mentioned and it fails deterministically in the previous implementation

```
    @Test
    public void testConcurrentAccessDoesNotLosePreviousState() {
        long maxSnapshotAgeMs = 10L;
        CountDownLatch latch = new CountDownLatch(1);
        Recorder mockRecorder = new MockRecorder(latch);
        HdrHistogram hdrHistogram = new HdrHistogram(maxSnapshotAgeMs, MAX_VALUE, 3, mockRecorder);

        AtomicReference<org.HdrHistogram.Histogram> asyncHistogram = new AtomicReference<>();
        org.HdrHistogram.Histogram mainHistogram;

        CompletableFuture<Void> asyncFuture = CompletableFuture.runAsync(() -> {
            asyncHistogram.set(hdrHistogram.latestHistogram(11));
        });

        mainHistogram = hdrHistogram.latestHistogram(11);

        asyncFuture.join();
        assertNotSame(asyncHistogram.get(), mainHistogram);
    }

    private static class MockRecorder extends Recorder {
        private final CountDownLatch latch;

        private final AtomicInteger invocationCounter = new AtomicInteger(0);

        public MockRecorder(CountDownLatch latch) {
            super(3);
            this.latch = latch;
        }

        @Override
        public org.HdrHistogram.Histogram getIntervalHistogram() {
            if (invocationCounter.compareAndSet(0, 1)) {
                try {
                    latch.await();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            } else {
                latch.countDown();
            }
            return super.getIntervalHistogram();
        }
    }
```

This test hangs in the new implementation because the call to `getIntervalHistogram` is synchronized on `lock`. Which makes sense since we want to ensure that the thread that calls getIntervalHistogram is the only thread that can proceed (and update the state). I wonder if we can modify this test to pass when it hangs for x seconds. What do you think?

Reply:
I made some improvements to the test I mentioned [here](https://github.com/apache/kafka/pull/17184#issuecomment-2347304340), and as a result it now runs much faster (well under a second on my development machine) and fails reliably without this change (failed 500 out of 500 attempts locally when run via Gradle).

After these improvements I prefer the aforementioned test over the one here, as the latter requires changes in the histogram code, is still not deterministically proving things if we are going to make it pass on a timeout, and is a bit harder to understand.